The best-fit allocator must prepare the heap for compaction. Fixes #9736

damiendoligez opened the pull request:
The remnants that are inserted at the head of the small free lists by the best-fit allocator, are free blocks with white color, which breaks the invariants of the compactor. The fix is to turn them blue before the compaction starts. That breaks the free-list invariants (ordering of blocks in the lists) but it doesn't matter because the compactor will rebuild the free lists anyway.

cc @lpw25

damiendoligez committed:
The best-fit allocator must prepare the heap for compaction. Fixes #9736

lpw25 commented:
The fix looks good.

With this change there is some freelist code you run before a compaction: the new `setup_for_compaction` function, and some freelist code you run after a compaction: the existing `reset` function. The need for two functions makes sense because between those two calls you may change allocation policy. However, the division of labour between them is maybe a bit odd. I would expect the call before the compaction to clear all the structures in the free list, and probably only the `init_merge` function to happen after the compaction.

So a possible improvement would be to put the new code in the `reset` function, call the `reset` function before compaction, remove the calls to `init_merge` from the reset functions, and call `init_merge` where we currently call `reset`.

damiendoligez commented:
That makes sense although we have to be careful around the policy changes.

lpw25 reviewed (approved):
LGTM

nojb commented:
This needs to be cherry-picked to 4.10 and 4.11 as well, no?

lpw25 commented:
I think it probably should, although I saw that Damien put the Changes entry in the 4.12 section. Up to @Octachron I guess.

Octachron commented:
Since we want users to *test* the best-fit allocator in the best possible conditions, it seems sensible to me to integrate those bug fixes in a last 4.11 beta and 4.10.1 . @damiendoligez , do you agree with the assessment or do you feel that backporting this specific bug fix would be too adventurous?

damiendoligez commented:
It doesn't seem too risky, although this is a very low-frequency bug so we won't get a lot of feedback anyway.

I'll do the cherry-picking.

damiendoligez commented:
Cherry-picked to 4.10 (54294fa7b0e0ba98a668bb8d53e14451990d9745)

Cherry-picked to 4.11 (b04908bdb77f6992328674db8d4c4c3288495c6e)

Moved and fixed the Changes entry in trunk (e509c796f5fba4bd5aba76819867a5836d891ad8)
